packages/web-workers/src/types.d.ts

```diff
@@ -0,0 +1,17 @@
+export function createWorker<T>(load: () => Promise<T>): () => Promisify<T>;
+
+type Arguments<T> = T extends (...arg: infer U) => any ? U : any;
+
+export type Promisify<T> = T extends (...args: infer Args) => infer TypeReturned
+  ? TypeReturned extends () => infer Result
+    ? (...args: Args) => () => Promise<Result>
+    : TypeReturned extends Promise<any> // check to prevent returning Promise<Promise<thing>>> in async function
+      ? (...args: Args) => TypeReturned
+      : (...args: Args) => Promise<TypeReturned>
```

Comment:
I don't think I understand this entire block (from `TypeReturned extends () => ...` down). Shouldn't this type just be applied recursively on the return type (so `() => {foo(): string}` becomes `Promise<() => Promise<{foo(): Promise<string>}>>`)

Reply:
This is because of the return types on promisey/async function cases. 

The snippet below is typed as `(() => string | Promise<string>): string | Promise<string>`

```
export function functionArgStringReturn(
  inputFunction: WorkerInput<() => string>,
) {
  const result = inputFunction();
  return typeof result === 'string'
    ? `Hello ${result}`
    : result.then(final => `Hello ${final}`);
}
```

The return type is `string | Promise<string>`. When we augment this type with `Promisify` it gets augmented to `Promise<string> | Promise<Promise<string>>`. This conditional type checks against this `Promise<Promise<string>>` case because we don't want to return the api as `Promise<Promise<string>>`